RecoMuon/MuonIdentification/src/MuonShowerInformationFiller.cc

```diff
@@ -838,28 +809,83 @@ void MuonShowerInformationFiller::fillHitsByStation(const reco::Muon& muon) {
        << theAllStationHits.at(3) << endl;
 
   //station shower sizes
-  MuonTransientTrackingRecHit::MuonRecHitContainer muonRecHitsPhiTemp, muonRecHitsPhiBest;
+  MuonTransientTrackingRecHit::MuonRecHitContainer muonRecHitsPhiBest;
   TransientTrackingRecHit::ConstRecHitContainer muonRecHitsThetaTemp, muonRecHitsThetaBest;
-
   // send station hits to the clustering algorithm
   for ( int stat = 0; stat != 4; stat++ ) {
-    if (!muonRecHits[stat].empty()) {
-      stable_sort(muonRecHits[stat].begin(), muonRecHits[stat].end(), LessPhi());
-
-      float dphimax = 0;
-      for (MuonTransientTrackingRecHit::MuonRecHitContainer::const_iterator iseed = muonRecHits[stat].begin(); iseed != muonRecHits[stat].end(); ++iseed) {
-          if (!(*iseed)->isValid()) continue;
-          GlobalPoint refpoint = (*iseed)->globalPosition(); //starting from the one with smallest value of phi
-          muonRecHitsPhiTemp.clear();
-          muonRecHitsPhiTemp = findPhiCluster(muonRecHits[stat], refpoint); //get clustered hits for this iseed
-      if (muonRecHitsPhiTemp.size() > 1) {
-         float dphi = fabs(deltaPhi((float)muonRecHitsPhiTemp.back()->globalPosition().barePhi(), (float)muonRecHitsPhiTemp.front()->globalPosition().barePhi()));
-         if (dphi > dphimax) {
-            dphimax = dphi;
-            muonRecHitsPhiBest = muonRecHitsPhiTemp;
-          }
-        } //at least two hits
-      }//loop over seeds 
+    const size_t nhit = muonRecHits[stat].size();
+    if (nhit < 2) continue; // Require at least 2 hits
+    muonRecHitsPhiBest.clear();
+    muonRecHitsPhiBest.reserve(nhit);
+
+    // Cluster seeds by global position phi. Best cluster is chosen to give greatest dphi
+    // Sort by phi (complexity = NLogN with enough memory, or = NLog^2N for insufficient mem)
+    stable_sort(muonRecHits[stat].begin(), muonRecHits[stat].end(), LessPhi());
+
+    // Search for gaps (complexity = N)
+    std::vector<size_t> clUppers;
+    for ( size_t ihit = 0; ihit < nhit; ++ihit ) {
+      const size_t jhit = (ihit+1)%nhit;
+      const double phi1 = muonRecHits[stat].at(ihit)->globalPosition().phi();

```

Comment:
.. implies to do it everywhere in the file: 's/globalPosition().phi()/globalPosition().barePhi()/'